app/code/Magento/CatalogInventory/Model/ResourceModel/Product/StockStatusBaseSelectProcessor.php

```diff
@@ -56,7 +56,9 @@ public function process(Select $select)
                 ['stock' => $stockStatusTable],
                 sprintf('stock.product_id = %s.entity_id', BaseSelectProcessorInterface::PRODUCT_TABLE_ALIAS),
                 []
-            )->where('stock.stock_status = ?', Stock::STOCK_IN_STOCK);
+            )
+                ->where('stock.stock_status = ?', Stock::STOCK_IN_STOCK)
+                ->where('stock.website_id = ?', 0);
```

Comment:
I don't see a lot of potential problems if we would accept proposed changes, but currently, Magento 2 is a Single Stock system and if your installation is not customized one, there should not be another data in `cataloginventory_stock_item` and `cataloginventory_stock_status` tables except data for Default Website (0). Thus there are no possibilities to have in_stock data on one website for particular SKU, and out of stock data for the same SKU on another website. That's functionality would be provided by Multi-Source Inventory (MSI) project.

All we do now is storing and retrieving data for Default Website only.
That's why most of the code looks like 
```
    /**
     * Subtract product qtys from stock.
     * Return array of items that require full save
     *
     * @param string[] $items
     * @param int $websiteId
     * @return StockItemInterface[]
     * @throws \Magento\Framework\Exception\LocalizedException
     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
     */
    public function registerProductsSale($items, $websiteId = null)
    {
        //if (!$websiteId) {
            $websiteId = $this->stockConfiguration->getDefaultScopeId();
        //}
```
So, I suppose that there is some customization on Magento Installation where current fix need to be applied.

But that fix could be also achieved pretty easily using Magento extensibility, as the class mentioned in this PR
```
class StockStatusBaseSelectProcessor implements BaseSelectProcessorInterface
```
being the implementation of Extension point declared in Magento (BaseSelectProcessorInterface).

Thus even in current code base, there is a possibility to achieve proposed behavior.